src/java/org/apache/nutch/crawl/Generator.java

```diff
@@ -114,81 +122,34 @@ public String toString() {
-      maxCount = job.getInt(GENERATOR_MAX_COUNT, -1);
-      if (maxCount == -1) {
-        byDomain = false;
-      }
-      if (GENERATOR_COUNT_VALUE_DOMAIN.equals(job.get(GENERATOR_COUNT_MODE)))
-        byDomain = true;
-      filters = new URLFilters(job);
-      normalise = job.getBoolean(GENERATOR_NORMALISE, true);
-      if (normalise)
-        normalizers = new URLNormalizers(job,
-            URLNormalizers.SCOPE_GENERATE_HOST_COUNT);
-      scfilters = new ScoringFilters(job);
-      partitioner.configure(job);
-      filter = job.getBoolean(GENERATOR_FILTER, true);
-      genDelay = job.getLong(GENERATOR_DELAY, 7L) * 3600L * 24L * 1000L;
-      long time = job.getLong(Nutch.GENERATE_TIME_KEY, 0L);
-      if (time > 0)
-        genTime.set(time);
-      schedule = FetchScheduleFactory.getFetchSchedule(job);
-      scoreThreshold = job.getFloat(GENERATOR_MIN_SCORE, Float.NaN);
-      intervalThreshold = job.getInt(GENERATOR_MIN_INTERVAL, -1);
-      restrictStatus = job.get(GENERATOR_RESTRICT_STATUS, null);
-      expr = JexlUtil.parseExpression(job.get(GENERATOR_EXPR, null));
-      maxNumSegments = job.getInt(GENERATOR_MAX_NUM_SEGMENTS, 1);
-      segCounts = new int[maxNumSegments];
-      
-      if (job.get(GENERATOR_HOSTDB) != null) {
-        try {
-          Path path = new Path(job.get(GENERATOR_HOSTDB), "current");
-          hostdbReaders = SequenceFileOutputFormat.getReaders(job, path);
-          maxCountExpr = JexlUtil.parseExpression(job.get(GENERATOR_MAX_COUNT_EXPR, null));
-          fetchDelayExpr = JexlUtil.parseExpression(job.get(GENERATOR_FETCH_DELAY_EXPR, null));
-        } catch (IOException e) {
-          LOG.error("Error reading HostDB because {}", e.getMessage());
-        }
-      }
-    }
+  public static class Selector extends
+      Partitioner<FloatWritable, Writable> {
+    private static LongWritable genTime = new LongWritable(System.currentTimeMillis());
+    private static long curTime;
+    private static long limit;
+    private static int segCounts[];
+    private static int maxCount;
+    private static boolean byDomain = false;
+    private static URLFilters filters;
+    private static URLNormalizers normalizers;
+    private static ScoringFilters scfilters;
+    private static SelectorEntry entry = new SelectorEntry();
+    private static FloatWritable sortValue = new FloatWritable();
+    private static boolean filter;
+    private static boolean normalise;
+    private static long genDelay;
+    private static FetchSchedule schedule;
+    private static float scoreThreshold = 0f;
+    private static int intervalThreshold = -1;
+    private static String restrictStatus = null;
+    private static int maxNumSegments = 1;
+    private static Expression expr = null;
+    private static MapFile.Reader[] hostdbReaders = null;
+    private static Expression maxCountExpr = null;
+    private static Expression fetchDelayExpr = null;
+    private static Configuration config;
+
```

Comment:
@sebastian-nagel I will look at the usage of each variable and declare them in their respective mapper and reducer classes as class variables, which are not obligated to be static. Does that sound good?

Reply:
Yes, that seems clear and safe. Of course, not to forget about setting the values from the configuration.

Reply:
>>Sharing configuration values between mapper and reducer via static variables does not work in distributed mode because mapper and reducer running each in its own JavaVM

@sebastian-nagel regarding this, if in distributed mode mapper and reducer are running on each of its own JVM how do they share the context object? 

>>Of course, not to forget about setting the values from the configuration.

and regarding this, we set values in configuration object before running the job itself and then they are later on used by mapper and reducer to retrieve the values right? And moreover, if the mapper or reducer are setting any value in configuration object I will make sure that it is happening in setup() respectively. Please feel free to correct me if I am missing something.

Reply:
Strictly speaking, it's a different object but it represents the "same" "job context" - configuration, counters, output location, etc.

You can set properties in the configuration object in the run() method of a job. The changed configuration is then passed via setup(Context) to the mapper and reducer objects. But it does not work the other way: if the configuration is modified in the setup() method, this has no effects to other mapper or reducer instances. The setup(context) method is called only once at the beginning of a task which makes it the right place to initialize variables from the configuration, create and configure required objects and plugins.